cocos/core/pipeline/multiLightDSManager.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ Copyright (c) 2020 Xiamen Yaji Software Co., Ltd.
+
+ https://www.cocos.com/
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated engine source code (the "Software"), a limited,
+ worldwide, royalty-free, non-assignable, revocable and non-exclusive license
+ to use Cocos Creator solely to develop games on your target platforms. You shall
+ not use Cocos Creator software for developing other software or tools that's
+ used for developing games. You are not granted to publish, distribute,
+ sublicense, and/or sell copies of Cocos Creator.
+
+ The software or tools in this License Agreement are licensed, not sold.
+ Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ */
+
+/**
+ * @packageDocumentation
+ * @module pipeline
+ */
+
+import { RenderPipeline } from './render-pipeline';
+import { Device, BufferUsageBit, MemoryUsageBit, BufferInfo,
+    Filter, Address, Sampler, DescriptorSet, DescriptorSetInfo } from '../gfx';
+import { Light } from '../renderer/scene/light';
+import { UBOGlobal, UBOShadow, UBOCamera, UNIFORM_SHADOWMAP_BINDING,
+    UNIFORM_SPOT_LIGHTING_MAP_TEXTURE_BINDING, PipelineGlobalBindings } from './define';
+import { genSamplerHash, samplerLib } from '../renderer/core/sampler-lib';
+import { builtinResMgr } from '../builtin/builtin-res-mgr';
+import { Texture2D } from '../assets/texture-2d';
+
+const _samplerInfo = [
+    Filter.LINEAR,
+    Filter.LINEAR,
+    Filter.NONE,
+    Address.CLAMP,
+    Address.CLAMP,
+    Address.CLAMP,
+];
+
+export class MultiLightDSManager {
+    private _pipeline: RenderPipeline;
+    private _device: Device;
+    private _descriptorSetMap: Map<Light, DescriptorSet> = new Map();
+    private _sampler: Sampler;
+
+    constructor (pipeline: RenderPipeline) {
+        this._pipeline = pipeline;
+        this._device = pipeline.device;
+
+        const shadowMapSamplerHash = genSamplerHash(_samplerInfo);
+        this._sampler = samplerLib.getSampler(this._device, shadowMapSamplerHash);
+    }
+
+    public updateDescriptorSet (light: Light) {
+        const descriptorSet = this._descriptorSetMap.get(light);
+        if (descriptorSet) {
+            const globalDescriptorSet = this._pipeline.descriptorSet;
+            for (let i = PipelineGlobalBindings.SAMPLER_SHADOWMAP; i < PipelineGlobalBindings.COUNT; i++) {
```

Comment:
this loop assumes `PipelineGlobalBindings.SAMPLER_SHADOWMAP` is the first sampler binding. Maybe put a comment near the binding declaration to make people aware of this when changing binding orders.

Reply:
sure